data/blog/umdctf-2023/pokeptcha.md

```diff
@@ -0,0 +1,414 @@
+function(){
+    var a = stack.pop();
+    var b = stack.pop();
+    stack.push(a[b])
+}
+
+// 18, add
+function(){
+    var a = stack.pop();
+    var b = stack.pop();
+    stack.push(a+b)
+}
+
+// 19, shr
+function(){
+    var a = stack.pop();
+    var b = stack.pop();
+    stack.push(b>>a)
+}
+
+// 20, jne again?
+function(){
+    var a = stack.pop();
+    var b = stack.pop();
+    var dst = stack.pop();
+    pc = a != b ? dst : pc
+}
+```
+
+Not quite sure about the two `jne` instructions, could be that I switched some variables around and one is actually a `je`, but from now I just printed the operations and didn’t interpret them anymore, so it didn’t matter too much (I just needed to know that there is a conditional jump to some address). With a bit of cleanup I got:
+
+```
+mem[99] = globalThis || globalThis.window
+[...]
+mem[1] = []
+mem[1].append(b'var LtTZ7p=function(){debugger};LtTZ7p();LtTZ7p')
+mem[1].append(5)
+setInterval(null, mem[1])
+mem[20] = mem[99].location.hostname
+mem[51] = FQgZw // the name of the whole interpreter function
+mem[20] += mem[51].toString(mem[51], [])
+mem[63] = mem[99].Date
+mem[20] += (mem[63].now(mem[63], []) - QZOR) >> 3 // time diff
+mem[87] = []
+mem[87].append(mem[20])
+mem[20] = md5(null, mem[87]) // md5(<whole string>)
+[...]
+```
+
+So we can see that we’re first calling setInterval with a debugger statement every 5 milliseconds. Then we do some string concatenations and finally call MD5 (which returns a hex string).
+
+Continuing we got:
+
+```
+mem[21] = r1
+mem[1] = []
+mem[2] = 0
+mem[3] = 0
+mem[4] = []
+
+// rc4 initial state [i for i in range(256)]
+mem[10] = 0
+loc_4388:
+	mem[1][mem[10]] = mem[10]
+	mem[10] = mem[10] + 1
+jmp loc_4388 if 256 != mem[10]
+
+// rc4 key init
+mem[10] = 0
+loc_4431:
+	mem[2] = mem[1][mem[10]] + mem[2]
+	mem[6] = mem[10] % mem[20].length
+	mem[5] = []
+	mem[5].append(mem[6])
+	mem[2] = mem[20].charCodeAt(mem[20], mem[5]) + mem[2]
+	mem[2] = mem[2] % 256
+	mem[3] = mem[1][mem[10]]
+	mem[1][mem[10]] = mem[1][mem[2]]
+	mem[1][mem[2]] = mem[3]
+	mem[10] = mem[10] + 1
+	jmp loc_4431 if 256 != mem[10]
+
+mem[10] = 0
+mem[2] = 0
+mem[5] = 0
+loc_4622:
+	mem[10] = (mem[10] + 1) % 256
+	mem[2] = (mem[1][mem[10]] + mem[2]) % 256
+	mem[3] = mem[1][mem[10]]
+	mem[1][mem[10]] = mem[1][mem[2]]
+	mem[1][mem[2]] = mem[3]
+	mem[7] = []
+	mem[7].append(mem[5])
+	mem[8] = mem[21].charCodeAt(mem[21], mem[7]) ^ mem[1][(mem[1][mem[10]] + mem[1][mem[2]])%256]
+	mem[7] = []
+	mem[7].append(mem[8])
+	mem[25] = String.fromCharCode(null, mem[7])
+	mem[4] = mem[25] + mem[4]
+	mem[5] = mem[5] + 1
+	jmp loc_4622 if mem[21].length != mem[5]
+```
+
+So it uses the MD5 hex-string as RC4 key, to then encrypt (or decrypt) our input which means we can get the RC4 key (Either using the JS debugger console and the convenient MD5 function or by copying everything into a Python script to calculate it there). Finally we just need to know the ciphertext to decrypt it. So more bytecode:
+
+```
```

Comment:
```suggestion
```js
```

Reply:
These are disassembled bytecode text and we need to use raw text instead of js.

Reply:
Even if it's bytecode, it's still significantly easier to read than raw JS. The syntax highlighter is unaffected.